cli/crates/common/src/environment.rs

```diff
@@ -330,6 +334,104 @@ fn find_grafbase_configuration(path: &Path, warnings: &mut Vec<Warning>) -> Opti
     }
 }
 
+#[derive(Clone, Copy, Debug)]
+pub enum NodePackageManager {
+    Npm,
+    Pnpm,
+    Yarn,
+}
+
+impl Default for NodePackageManager {
+    fn default() -> Self {
+        Self::Npm
+    }
+}
+
+impl NodePackageManager {
+    #[must_use]
+    pub fn is_npm(self) -> bool {
+        matches!(self, Self::Npm)
+    }
+}
+
+impl fmt::Display for NodePackageManager {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Npm => f.write_str("npm"),
+            Self::Pnpm => f.write_str("pnpm"),
+            Self::Yarn => f.write_str("yarn"),
+        }
+    }
+}
+
+impl FromStr for NodePackageManager {
+    type Err = CommonError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "npm" => Ok(NodePackageManager::Npm),
+            "pnpm" => Ok(NodePackageManager::Pnpm),
+            "yarn" => Ok(NodePackageManager::Yarn),
+            _ => {
+                let inner = io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("Unsupported package manager in package.json: {s}"),
+                );
+
+                Err(CommonError::AccessPackageJson(inner))
+            }
+        }
+    }
+}
+
+pub fn install_dependencies(project_root: &Path) -> Result<(), CommonError> {
+    let command = match which::which("npm") {
+        Ok(command) => command,
+        Err(e) => return Err(CommonError::NpmNotFound(Box::new(e))),
+    };
+
+    Command::new(command)
+        .args(["install", "--save-dev"])
+        .current_dir(project_root)
+        .output()
+        .map_err(CommonError::NpmInstall)?;
+
+    Ok(())
+}
+
+pub fn guess_project_package_manager(project_root: &Path, package_json: &Path) -> Option<NodePackageManager> {
```

Comment:
Why the change in the implementation?
For one, this function gets called from the UDF build infrastructure which is async, thus effectively inserting calls blocking on I/O rather than yielding. It's not a tragedy, but a change for the worse.
Any particular reason for the rewrite from into a loop?
Also, we shouldn't use statics for cached computation. The `Project` structure might be a good place to keep the package manager cached if we want it cached.

Reply:
In general, I'd like to avoid async if we don't need async. In CLI's case, there's a tokio executor already with as many threads as there are cores. Having a blocking file system access is much nicer developer experience than playing with the pretty ugly apis from Tokio, especially when their APIs are not really using async IO for file system access, just running it in a separate thread. We're not in a JavaScript event loop, we have quite a many of those loops in Tokio.

For the complexity of async IO, is the CLI code needing that much of throughput? Will loading the package.json be blocking everything in every thread so it will affect throughput? Will we need to handle that many requests in the CLI that this will be an issue? Especially when we load the package manager to a cell, which is kept in cache after the first call?

For `Itertools`, in Prisma we got slapped when we tried to use that. It's a bit of a mess of a crate, especially when some of the iterators actually allocate which is not cool. I always prefer a simple for loop, if there's no need for composition.

You can't cache it in the `Project` structure due to it not existing on init where we need this. I'd need to change the whole project initialization a lot so I could first run init, and after that create a project. Much less friction this way.